### PR TITLE
Feature/traffic stats

### PR DESCRIPTION
**Pull request by gtosatti**
New feature to collect Traffic info (outer) a get the top N Flow (by num of packets)

### CODE REVIEW

File: src/timestamp_pkt_processor.h [hunk 1]
```diff
@@ -68,23 +68,23 @@ class TimestampPacketProcessor : public IPacketProcessor {
...
-    bool is_some_processing_active() const
+    virtual bool is_some_processing_active() const
     {
         return m_proc_mode != PROCMODE_NONE;
     }
 
     // to compute correctly the timestamps in --set-duration mode, we need 2
     // passes: first to find out how many packets are present in the PCAP and then
     // to actually alter timestamps:
-    virtual bool needs_2passes() const override
```
Comment (f18m) [inline src/timestamp_pkt_processor.h:79]: no need to remove "virtual" from needs_2passes()

File: src/process_file.cpp [hunk 2]
```diff
@@ -315,7 +315,7 @@ static bool process_pcap_handle(pcap_t* pcap_handle_in,
     }
 
     if (processorcfg) {
-        if (!processorcfg->post_processing(nloaded))
+        if (!processorcfg->post_processing(infile.c_str(), nloaded))
```
Comment (f18m) [inline src/process_file.cpp:318]: c_str() can be removed on all post_processing() invocation

File: src/large-pcap-analyzer.cpp [hunk 3]
```diff
@@ -106,6 +109,9 @@ static void print_help()
     printf(" -q,--quiet               suppress all normal output, be script-friendly\n");
     printf(" -t,--timing              provide timestamp analysis on loaded packets\n");
     printf(" -p,--stats               provide basic parsing statistics on loaded packets\n");
+    printf(" -x,--trafficstats<FlowNum>\n");
```
Comment (f18m) [inline src/large-pcap-analyzer.cpp:112]: I was thinking about instead providing a single --traffic-report option that takes a string which is the "name of the report" that you want to produce. 
E.g. for now we might have only 2 reports "top10flows_by_pkts" and "allflows_by_pkts" . In the future other reports might be added to the list of supported reports.

File: src/FastIpAddress.h [hunk 4]
```diff
@@ -0,0 +1,307 @@
+//------------------------------------------------------------------------------
```
Comment (f18m) [inline src/FastIpAddress.h:1]: let's reduce this to the minimal required portion of FastIpAddress class that is needed. Then maybe we can rename to simply "IpAddress" this whole class and "ipaddress.h" the file

File: src/IpAddress.h [hunk 5]
```diff
@@ -0,0 +1,352 @@
+//------------------------------------------------------------------------------
```
Comment (f18m) [inline src/IpAddress.h:1]: this can be entirely removed

File: src/SCTPdefs.h [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+/*
+ * Header repeated only once just after IP frame
+ */
+typedef struct sctphdr {
+    uint16_t source;
+    uint16_t dest;
+    uint32_t vtag;
+    uint32_t checksum;
+} __attribute__((packed)) sctp_sctphdr_t; // 12B
```
Comment (f18m) [inline src/SCTPdefs.h:36]: this "sctp_sctphdr_t" header is probably the only thing that is needed from this header right?

File: src/large-pcap-analyzer.cpp (hunk 3 above)
Comment (gtosatti) [reply src/large-pcap-analyzer.cpp]: Done. From my point of view the previous version was a bit easier... however, no problem. By the way, we also have a separate "inner" option to enable the trafficStats using the inner values. is it ok ?

Comment (f18m) [reply src/large-pcap-analyzer.cpp]: Well my idea is to allow different combinations of inner/outer using different report names e.g. 'top10flows_by_pkts_outer' and 'top10flows_by_pkts_inner'

There are too many variations to have 1 cli option for each of them...

**Comment by f18m** [comment]
note to myself:
a) still need to review trafficprocessor files
b) fix README.md
c) need to create a new test in run_tests.sh for the new --report option

**Comment by f18m** [comment]
@gtosatti do you already have some PCAP file (not containing sensitive data) that you already manually verified and we can commit to this repo to improve automated tests? I currently I have put only ipv4_ftp.pcap into the automatic tests...

**Comment by gtosatti** [comment]
> @gtosatti do you already have some PCAP file (not containing sensitive data) that you already manually verified and we can commit to this repo to improve automated tests? I currently I have put only ipv4_ftp.pcap into the automatic tests...

Not really, I just used the large Customer pcap file that I needed to analyze.

**Comment by gtosatti** [comment]
Question... not sure to understand, are you expecting any code change from my side ? Please confirm.

**Comment by f18m** [comment]
> Question... not sure to understand, are you expecting any code change from my side ? Please confirm.

hi @gtosatti , no code change, no.
But I think it would be really useful if you could test again the large-pcap-analyzer after my changes to address your original use case (i.e. re-analyze the large PCAP file you analyzed initially) and check that you can get the same CSV report.

thanks

**Comment by gtosatti** [comment]
> > Question... not sure to understand, are you expecting any code change from my side ? Please confirm.
> 
> hi @gtosatti , no code change, no. But I think it would be really useful if you could test again the large-pcap-analyzer after my changes to address your original use case (i.e. re-analyze the large PCAP file you analyzed initially) and check that you can get the same CSV report.
> 
> thanks

I used the original large pcap file with the latest large-pcap-analyzer version and I got the same result (inner and outer list). So, from my perspective, it is still working fine.

**Pull request merged**